kiva/markers/marker_renderer.h

```diff
@@ -0,0 +1,130 @@
+#ifndef KIVA_MARKER_RENDERER_H
+#define KIVA_MARKER_RENDERER_H
+
+#include <agg_pixfmt_rgb.h>
+#include <agg_pixfmt_rgba.h>
+#include <agg_renderer_base.h>
+#include <agg_renderer_markers.h>
+#include <agg_trans_affine.h>
+
+namespace kiva_markers
+{
+    // This enumeration must match the marker constants in `kiva.constants`!
+    enum marker_type
+    {
+        MARKER_SQUARE = 1,
+        MARKER_DIAMOND,
+        MARKER_CIRCLE,
+        MARKER_CROSSED_CIRCLE,
+        MARKER_CROSS,
+        MARKER_TRIANGLE,
+        MARKER_INVERTED_TRIANGLE,
+        MARKER_PLUS,
+        MARKER_DOT,
+        MARKER_PIXEL,
+    };
+
+    class marker_renderer_base
+    {
+    public:
+        virtual ~marker_renderer_base(){}
+        virtual bool draw_markers(const double* pts, const unsigned Npts,
+                                  const unsigned size, const marker_type type,
+                                  const double* fill, const double* stroke) = 0;
+        virtual void transform(const double sx, const double sy, const double shx,
+                               const double shy, const double tx, const double ty) = 0;
+    };
+
+    template<typename pixfmt_t>
+    class marker_renderer : public marker_renderer_base
+    {
+    public:
+        marker_renderer(unsigned char* buf,
+                        const unsigned width, const unsigned height,
+                        const int stride, const bool bottom_up = false)
+        : m_renbuf(buf, width, height, bottom_up ? -stride : stride)
+        , m_pixfmt(m_renbuf)
+        , m_base_renderer(m_pixfmt)
+        , m_renderer(m_base_renderer)
+        {}
+
+        virtual ~marker_renderer() {}
+
+        bool draw_markers(const double* pts, const unsigned Npts,
+                          const unsigned size, const marker_type type,
+                          const double* fill, const double* stroke)
+        {
+            // Map from our marker type to the AGG marker type
+            const agg24markers::marker_e marker = _get_marker_type(type);
+            if (marker == agg24markers::end_of_markers) return false;
+
+            // Assign fill and line colors
+            m_renderer.fill_color(agg24markers::rgba(fill[0], fill[1], fill[2], fill[3]));
+            m_renderer.line_color(agg24markers::rgba(stroke[0], stroke[1], stroke[2], stroke[3]));
+
+            // NOTE: this is the average in X and Y
+            const double scale = m_transform.scale();
+
+            // Draw the markers
+            double mx, my;
+            for (unsigned i = 0; i < Npts*2; i+=2)
+            {
+                mx = pts[i];
+                my = pts[i+1];
+                m_transform.transform(&mx, &my);
+                m_renderer.marker(int(mx), int(my), size * scale, marker);
+            }
+
+            return true;
+        }
+
+        void transform(const double sx, const double sy,
+                       const double shy, const double shx,
+                       const double tx, const double ty)
```

Comment:
i'm sorry but `transform` is freaking me out now - the signature for `transform` on `marker_renderer_base` does not match this - again, the `shy` and `shx` are switched around. I apologize if im missing something obvious.

Reply:
I'm glad you're looking close enough to spot it. Hopefully my latest commit has it right now. 😅